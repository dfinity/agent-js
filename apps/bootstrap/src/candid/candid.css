:root #candid_ui {
  font-family: 'Circular Std', 'Roboto', sans-serif;
  font-weight: 300;
  font-size: 10px;
  --font-circular: 'Circular Std', 'Roboto', sans-serif;
  --font-mono: 'Vremena Grotesk', 'Roboto Mono', monospace;
  --header-height: 3.8rem;
  --pad-sm: 0.8rem;
  --pad-md: 1.5rem;
  --pad-lg: 2.4rem;
  --font-sm: 1.2rem;
  --font-md: 1.4rem;
  --font-lg: 2.4rem;
  --light-bg: #f6f6f7;
  --white: #f9f9f9;
  --lightest: #ebf0fa;
  --lighter: #eaeaea;
  --light: #e5e5e5;
  --medium: #c3c3c4;
  --dark: #818284;
  --darker: #545454;
  --darkest: #292a2e;
  --error: #b82121;
}

#candid_ui #app {
  display: flex;
  flex-direction: column;
  width: 100vw;
  height: 100vh;
  font-size: var(--font-md);
}

#candid_ui #container {
  display: flex;
  flex-direction: column;
  height: calc(100vh - var(--header-height));
}

#candid_ui #main-content {
  flex: 1;
  overflow: auto;
}

#candid_ui #header {
  height: var(--header-height);
  width: 100%;
  text-align: center;
  font-size: var(--font-md);
  line-height: var(--header-height);
  color: var(--darkest);
  border-bottom: 1px solid var(--dark);
}

#candid_ui #title-card {
  padding: 3.2rem 1.8rem 0;
  font-size: 1.6rem;
}

#candid_ui #title {
  font-size: 3.4rem;
  letter-spacing: 0.1rem;
}

#candid_ui #methods {
  font-family: 'Vremena Grotesk', 'Roboto Mono', monospace;
  padding: 0;
  margin: 0;
}

#candid_ui #methods li {
  display: flex;
  flex-direction: column;
  list-style: none;
  padding: var(--pad-md);
  margin-top: var(--pad-lg);
  border: 1px solid var(--dark);
}

#candid_ui .signature {
  font-weight: lighter;
  margin-bottom: var(--pad-md);
}
.#candid_ui signature b {
  font-weight: bold;
}

#candid_ui .composite {
  margin-bottom: var(--pad-md);
}

#candid_ui input,
#candid_ui select {
  font-size: var(--font-md);
  font-family: 'Circular Std', 'Roboto', sans-serif;
  margin-bottom: var(--pad-lg);
}
#candid_ui input:not([type='checkbox']),
#candid_ui select {
  width: 100%;
  height: 4.8rem;
  padding: 1.5rem;
  border: 1px solid var(--dark);
}
#candid_ui input[type='number']::-webkit-inner-spin-button {
  width: 3rem;
  height: 3rem;
}
#candid_ui label {
  font-family: 'Circular Std', 'Roboto', sans-serif;
  margin-right: var(--pad-sm);
}
#candid_ui label.small {
  font-size: 1rem;
}
#candid_ui .popup-form {
  padding-left: var(--pad-md);
  width: 100%;
}
#candid_ui input[type='checkbox'] ~ .popup-form {
  padding-left: var(--pad-md);
}
#candid_ui .input-container > span > .popup-form,
#candid_ui .ui-result > span > .popup-form {
  padding-left: 0;
}
#candid_ui .input-container span:not(.status),
#candid_ui .ui-result span:not(.status) {
  position: relative;
  overflow-clip: none;
  overflow-y: visible;
}

#candid_ui .reject {
  border-color: var(--error) !important;
  box-shadow: 0 0 1px var(--error);
}
#candid_ui .result {
  display: none;
  position: relative;
  width: 100%;
  margin-top: var(--pad-lg);
  padding-top: 1.5rem;
  font-size: 1.4rem;
  border-top: 1px solid var(--light);
}
#candid_ui .error {
  color: var(--error);
}
#candid_ui .status {
  display: none;
  position: absolute;
  top: 3.1rem;
  left: 0.2rem;
  font-size: 1.2rem;
  color: var(--error);
  white-space: pre;
}
#candid_ui .left {
  flex: 1;
  overflow-wrap: anywhere;
}
#candid_ui .right {
  color: var(--dark);
  align-self: stretch;
  margin-left: var(--pad-md);
}
#candid_ui .clear {
  clear: both;
}

#candid_ui .button-container {
  display: flex;
  margin-top: 0.8rem;
}

#candid_ui .btn {
  font-family: 'Roboto', sans-serif;
  letter-spacing: 0.05rem;
  background-color: var(--darkest);
  color: var(--lightest);
  font-size: 1.4rem;
  padding: 1.5rem;
  border: none;
  text-transform: uppercase;
  flex: 1;
}
#candid_ui .btn:not(:last-child) {
  margin-right: var(--pad-md);
}
#candid_ui button {
  padding: 0;
  border: none;
  background-color: inherit;
  color: inherit;
}
#candid_ui button:focus,
#candid_ui input:focus,
#candid_ui select:focus,
#candid_ui textarea:focus {
  outline: none;
}
#candid_ui .random {
  background-color: var(--lighter);
  color: var(--darkest);
}

#candid_ui .result-buttons {
  display: none;
  position: absolute;
  top: -0.1rem;
  right: 0;
}
#candid_ui .result:not(.error):hover .result-buttons {
  display: block;
}
#candid_ui .result-buttons .btn {
  color: var(--darkest);
  background-color: var(--white);
  border: 1px solid var(--darkest);
  font-size: 1.1rem;
  padding: 0.2rem 0.8rem;
}
#candid_ui .result-buttons .btn:not(:last-child) {
  margin-right: 0.5rem;
}
#candid_ui .result-buttons .btn:focus,
#candid_ui .result-buttons .btn:hover {
  background-color: var(--light);
}
#candid_ui .result-buttons .btn:active {
  background-color: var(--lightest);
}
#candid_ui .result-buttons .btn.active {
  color: var(--white);
  background-color: var(--darkest);
}

#candid_ui #console {
  overflow: hidden;
  height: var(--header-height);
  background-color: var(--light-bg);
  color: var(--darkest);
  transition: height 0.2s ease-in-out;
}
#candid_ui #console.open {
  height: 40vh;
}

#candid_ui #console-bar {
  display: flex;
  height: var(--header-height);
  background-color: var(--white);
  padding: 0;
  border-bottom: 1px solid var(--light);
}
#candid_ui .console-header {
  display: flex;
  align-items: center;
  justify-content: center;
  width: 100%;
  height: var(--header-height);
  background-color: var(--light-bg);
  font-size: 2rem;
  font-family: var(--font-circular);
  text-align: center;
  text-transform: uppercase;
}
#candid_ui #output-button,
#candid_ui #methods-button {
  flex: 1;
  cursor: pointer;
  font-size: var(--font-lg);
  border-bottom: 1px solid var(--medium);
}
#candid_ui #output-button:hover,
#candid_ui #methods-button:hover,
#candid_ui #output-button:focus,
#candid_ui #methods-button:focus {
  background-color: var(--light);
}
#candid_ui .active-tab {
  background-color: var(--lighter);
}

#candid_ui #output-pane,
#candid_ui #methods-pane {
  font-family: var(--font-mono);
  height: 0;
  padding: 0 0 3rem;
  margin-right: calc(-1 * var(--pad-md));
}

#candid_ui #output-list,
#candid_ui #methods-list {
  list-style: none;
  padding: 0 0 var(--pad-lg);
  height: calc(40vh - (var(--header-height) * 2));
  overflow-y: auto;
}

#candid_ui .output-line {
  width: 100%;
  padding: 0.4rem 0.8rem;
  overflow-wrap: break-word;
}
#candid_ui .output-line:nth-child(2n-1) {
  background-color: var(--lighter);
}

#candid_ui #methods-list {
  padding-top: 0.2rem;
}
#candid_ui #methods-list a {
  display: block;
  padding: 0.7rem 1.5rem;
  color: var(--darkest);
}
#candid_ui #methods-list a:focus,
#candid_ui #methods-list a:hover {
  background-color: var(--darker);
  color: var(--white);
  text-decoration: none;
}
#candid_ui #methods-list a:active {
  background-color: var(--darkest);
}

@media only screen and (min-width: 800px) {
  :root #candid_ui {
    --header-height: 5.6rem;
    --console-width: clamp(28rem, 30vw, 40rem);
  }

  #candid_ui #header {
    text-align: left;
    padding-left: 3rem;
  }

  #candid_ui #container {
    flex-direction: row;
    overflow: auto;
  }

  #candid_ui #title-card {
    padding-left: 0;
  }

  #candid_ui #main-content {
    flex: 1;
    padding: 0 3rem 5rem;
  }

  #candid_ui .btn {
    width: auto;
    flex: 0;
    padding: var(--pad-sm) var(--pad-md);
  }

  #candid_ui .input-container span:not(:first-child):not(.status),
  #candid_ui .ui-result span:not(:first-child):not(.status) {
    margin-left: var(--pad-md);
  }
  #candid_ui input:not([type='checkbox']),
  #candid_ui select {
    width: 23.2rem;
    height: 4rem;
    padding: 1rem var(--pad-md);
  }

  #candid_ui #console {
    display: flex;
    flex-direction: row-reverse;
    width: var(--header-height);
    height: calc(100vh - var(--header-height));
    padding: 0;
    transition: width 0.2s ease-in-out;
  }
  #candid_ui #console.open {
    width: calc(var(--header-height) + var(--console-width));
    height: calc(100vh - var(--header-height));
  }

  #candid_ui #console-bar {
    display: flex;
    flex-direction: column;
    justify-self: end;
    width: var(--header-height);
    height: 100%;
    padding: 0;
    border-bottom: none;
    border-left: 1px solid var(--light);
  }
  #candid_ui #console-bar button {
    flex: 0;
    font-size: 4rem;
    padding: 0.8rem;
  }

  #candid_ui .console-header {
    display: none;
    width: 0;
    border-bottom: none;
    transition: width 0.2s ease-in-out;
  }
  #candid_ui #console.open .console-header {
    display: flex;
    width: 100%;
  }

  #candid_ui #output-pane,
  #candid_ui #methods-pane {
    width: 100%;
    margin: 0;
  }

  #candid_ui #methods-pane,
  #candid_ui #output-pane {
    height: 100% !important;
    width: 0;
    transition: width 0.2s ease-in-out;
  }
  #candid_ui #console.open #methods-pane,
  #candid_ui #console.open #output-pane {
    width: var(--console-width);
  }

  #candid_ui #output-list,
  #candid_ui #methods-list {
    height: calc(100vh - (var(--header-height) * 2));
  }
}
