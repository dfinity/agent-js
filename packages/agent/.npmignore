# We work with a safelist here, so block everything that's not permitted, and add packages
# that are.
**

!package.json
!README.md
!LICENSE

!src/**/*.ts
src/**/*.test.ts
src/**/__certificates__
src/package.json

!lib/**
lib/**/*.test.*
lib/**/*.tsbuildinfo
 