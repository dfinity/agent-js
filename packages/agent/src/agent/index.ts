import { GlobalInternetComputer } from '../index';
import { Agent } from './api';

export * from './api';
export * from './http';
export * from './proxy';

declare const window: GlobalInternetComputer;
declare const global: GlobalInternetComputer;
declare const self: GlobalInternetComputer;

export function getDefaultAgent(): Agent {
  const agent =
    typeof window === 'undefined'
      ? typeof global === 'undefined'
        ? typeof self === 'undefined'
          ? undefined
          : self.ic.agent
        : global.ic.agent
      : window.ic.agent;

  if (!agent) {
    throw new Error('No Agent could be found.');
  }

  return agent;
}
