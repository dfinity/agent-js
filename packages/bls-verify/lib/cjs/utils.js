'use strict';
Object.defineProperty(exports, '__esModule', { value: true });
exports.compare = exports.fromHex = exports.toHex = exports.concat = void 0;
/**
 * Concatenate multiple array buffers.
 * @param buffers The buffers to concatenate.
 */
function concat(...buffers) {
  const result = new Uint8Array(buffers.reduce((acc, curr) => acc + curr.byteLength, 0));
  let index = 0;
  for (const b of buffers) {
    result.set(new Uint8Array(b), index);
    index += b.byteLength;
  }
  return result.buffer;
}
exports.concat = concat;
/**
 * Transforms a buffer to an hexadecimal string. This will use the buffer as an Uint8Array.
 * @param buffer The buffer to return the hexadecimal string of.
 */
function toHex(buffer) {
  return [...new Uint8Array(buffer)].map(x => x.toString(16).padStart(2, '0')).join('');
}
exports.toHex = toHex;
const hexRe = new RegExp(/^([0-9A-F]{2})*$/i);
/**
 * Transforms a hexadecimal string into an array buffer.
 * @param hex The hexadecimal string to use.
 */
function fromHex(hex) {
  if (!hexRe.test(hex)) {
    throw new Error('Invalid hexadecimal string.');
  }
  const buffer = [...hex]
    .reduce((acc, curr, i) => {
      // tslint:disable-next-line:no-bitwise
      acc[(i / 2) | 0] = (acc[(i / 2) | 0] || '') + curr;
      return acc;
    }, [])
    .map(x => Number.parseInt(x, 16));
  return new Uint8Array(buffer).buffer;
}
exports.fromHex = fromHex;
function compare(b1, b2) {
  if (b1.byteLength !== b2.byteLength) {
    return b1.byteLength - b2.byteLength;
  }
  const u1 = new Uint8Array(b1);
  const u2 = new Uint8Array(b2);
  for (let i = 0; i < u1.length; i++) {
    if (u1[i] !== u2[i]) {
      return u1[i] - u2[i];
    }
  }
  return 0;
}
exports.compare = compare;
//# sourceMappingURL=utils.js.map
