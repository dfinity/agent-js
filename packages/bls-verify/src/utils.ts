/**
 * Transforms a buffer to an hexadecimal string. This will use the buffer as an Uint8Array.
 * @param buffer The buffer to return the hexadecimal string of.
 */
export function toHex(buffer: ArrayBuffer): string {
  return [...new Uint8Array(buffer)].map(x => x.toString(16).padStart(2, '0')).join('');
}

const hexRe = new RegExp(/^([0-9A-F]{2})*$/i);

/**
 * Transforms a hexadecimal string into an array buffer.
 * @param hex The hexadecimal string to use.
 */
export function fromHex(hex: string): ArrayBuffer {
  if (!hexRe.test(hex)) {
    throw new Error('Invalid hexadecimal string.');
  }
  const buffer = [...hex]
    .reduce((acc, curr, i) => {
      acc[(i / 2) | 0] = (acc[(i / 2) | 0] || '') + curr;
      return acc;
    }, [] as string[])
    .map(x => Number.parseInt(x, 16));

  return new Uint8Array(buffer).buffer;
}
