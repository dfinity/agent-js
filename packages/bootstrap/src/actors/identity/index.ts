export { default as DocumentIdentities } from './DocumentIdentities';
export { default as IdentityActor } from './IdentityActor';
