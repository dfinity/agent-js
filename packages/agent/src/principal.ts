import { Buffer } from 'buffer/';
import { BinaryBlob, blobFromHex, blobFromUint8Array, blobToHex } from './types';
import { decode, encode } from './utils/base32';
import { getCrc32 } from './utils/getCrc';
import { sha224 } from './utils/sha224';

const SELF_AUTHENTICATING_SUFFIX = 2;
const ANONYMOUS_SUFFIX = 4;

export class Principal {
  public static anonymous(): Principal {
    return new this(blobFromUint8Array(new Uint8Array([ANONYMOUS_SUFFIX])));
  }

  public static selfAuthenticating(publicKey: BinaryBlob): Principal {
    const sha = sha224(publicKey);
    return new this(blobFromUint8Array(new Uint8Array([...sha, SELF_AUTHENTICATING_SUFFIX])));
  }

  public static from(other: unknown): Principal {
    if (typeof other === 'string') {
      return Principal.fromText(other);
    } else if (
      typeof other === 'object'
      && other !== null
      && (other as Principal)._isPrincipal === true
    ) {
      return new Principal((other as Principal)._blob);
    }

    throw new Error(`Impossible to convert ${JSON.stringify(other)} to Principal.`);
  }

  public static fromHex(hex: string): Principal {
    return new this(blobFromHex(hex));
  }

  public static fromText(text: string): Principal {
    const canisterIdNoDash = text.toLowerCase().replace(/-/g, '');

    let arr = decode(canisterIdNoDash);
    arr = arr.slice(4, arr.length);

    const principal = new this(blobFromUint8Array(arr));
    if (principal.toText() !== text) {
      throw new Error(`Principal "${principal.toText()}" does not have a valid checksum.`);
    }

    return principal;
  }

  public static fromBlob(blob: BinaryBlob): Principal {
    return new this(blob);
  }

  public readonly _isPrincipal = true;

  protected constructor(private _blob: BinaryBlob) {}

  public isAnonymous() {
    return this._blob.byteLength === 1 && this._blob[0] === ANONYMOUS_SUFFIX;
  }

  public toBlob(): BinaryBlob {
    return this._blob;
  }

  public toHash() {
    return this._blob;
  }

  public toHex(): string {
    return blobToHex(this._blob).toUpperCase();
  }

  public toText(): string {
    const checksumArrayBuf = new ArrayBuffer(4);
    const view = new DataView(checksumArrayBuf);
    view.setUint32(0, getCrc32(this.toBlob()), false);
    const checksum = Uint8Array.from(Buffer.from(checksumArrayBuf));

    const bytes = Uint8Array.from(this._blob);
    const array = new Uint8Array([...checksum, ...bytes]);

    const result = encode(array);
    const matches = result.match(/.{1,5}/g);
    if (!matches) {
      // This should only happen if there's no character, which is unreachable.
      throw new Error();
    }
    return matches.join('-');
  }

  public toString() {
    return this.toText();
  }
}
