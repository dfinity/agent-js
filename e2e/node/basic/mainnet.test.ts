import { Actor, AnonymousIdentity, HttpAgent, Identity, CanisterStatus } from '@dfinity/agent';
import { IDL } from '@dfinity/candid';
import { Ed25519KeyIdentity } from '@dfinity/identity';
import { Principal } from '@dfinity/principal';
import { describe, it, expect, vi } from 'vitest';
import { makeAgent } from '../utils/agent';

const createWhoamiActor = async (identity: Identity) => {
  const canisterId = 'ivcos-eqaaa-aaaab-qablq-cai';
  const idlFactory = () => {
    return IDL.Service({
      whoami: IDL.Func([], [IDL.Principal], ['query']),
    });
  };
  vi.useFakeTimers();
  new Date(Date.now());

  const agent = await makeAgent({ host: 'https://icp-api.io', identity });

  return Actor.createActor(idlFactory, {
    agent,
    canisterId,
  });
};

describe('certified query', () => {
  it('should verify a query certificate', async () => {
    const actor = await createWhoamiActor(new AnonymousIdentity());

    const result = await actor.whoami();
    expect(Principal.from(result)).toBeInstanceOf(Principal);
  }, 100_000);
  it('should verify lots of query certificates', async () => {
    let count = 0;
    const identities = Array.from({ length: 20 }).map(() => {
      const newIdentity = Ed25519KeyIdentity.generate(new Uint8Array(32).fill(count));
      count++;
      return newIdentity;
    });
    const actors = await identities.map(async identity => await createWhoamiActor(identity));

    const promises = actors.map(async actor => (await actor).whoami());

    const results = await Promise.all(promises);
    results.forEach(result => {
      expect(Principal.from(result)).toBeInstanceOf(Principal);
    });
    expect(results.length).toBe(20);

    expect(results).toMatchInlineSnapshot(`
      [
        {
          "__principal__": "535yc-uxytb-gfk7h-tny7p-vjkoe-i4krp-3qmcl-uqfgr-cpgej-yqtjq-rqe",
        },
        {
          "__principal__": "wf3fv-4c4nr-7ks2b-xa4u7-kf3no-32glf-lf7e4-4ng4a-wwtlu-a2vnq-nae",
        },
        {
          "__principal__": "52mr2-fw2ng-2ofst-7jekz-xbymo-3ysz7-itwdk-bgstz-r7g4g-oz5vi-pqe",
        },
        {
          "__principal__": "skpwg-42fe4-eyep5-nfyz7-66wvg-hthea-q3eek-vonbv-5wpxs-nxhmh-fqe",
        },
        {
          "__principal__": "ghaya-cncjm-ntxgt-af5pp-6hzsz-tvwlv-hrlfc-ocq3t-ai7vk-vyixr-cqe",
        },
        {
          "__principal__": "ebtho-zkeqd-ebs74-f5if3-lk6js-3a5q5-37xt4-3ylns-h7r4n-tkhly-aqe",
        },
        {
          "__principal__": "5zqn5-627q2-spx2w-mt6bm-llsnz-gipvv-5femn-3box4-vzuh6-bn723-sae",
        },
        {
          "__principal__": "tek7g-2zmny-nzjwg-ansf7-rkxv6-z32x6-3flbb-ous5d-pygjx-wkhlc-jae",
        },
        {
          "__principal__": "jjn6g-sh75l-r3cxb-wxrkl-frqld-6p6qq-d4ato-wske5-op7s5-n566f-bqe",
        },
        {
          "__principal__": "447dk-byguq-fqkfn-7h4r6-lpk74-itbnh-mkutj-m6tmy-igaff-hmfel-cqe",
        },
        {
          "__principal__": "muo3f-ines5-bxbwm-6wi5e-z663m-3zte2-r7d4x-pleey-xqvxt-scwc5-jae",
        },
        {
          "__principal__": "snzff-yj2qd-fjns7-lqhvw-rsgq7-tohk2-fjnw4-uq3d6-wtk56-pxgry-mqe",
        },
        {
          "__principal__": "pb54o-aqais-24v7j-msopl-bqeuv-paefp-vuoqc-gkezk-grujb-oetl6-sae",
        },
        {
          "__principal__": "bf37n-7ybos-wmqt6-yiov5-24q4m-ajpjk-madkr-snuj2-ngruk-tspnh-aqe",
        },
        {
          "__principal__": "tbjmy-vay5e-hqvv6-sval4-zfxmm-aii6f-b7p55-hwtz5-dejtl-qcuh2-aqe",
        },
        {
          "__principal__": "7d3pe-dh4ov-fp5xz-nctjc-5rduh-gzv3t-5ioyh-4dvx3-x4lgj-hz63t-dqe",
        },
        {
          "__principal__": "37axv-sazcg-75pi3-owhxr-kollq-xnzjz-zfxsv-nzdbp-yaelp-shcul-jae",
        },
        {
          "__principal__": "r772c-4dz5f-rpg4e-qzxgg-7bxlb-67zpu-bitgb-vsx7k-mmagd-6zk3d-4qe",
        },
        {
          "__principal__": "knkon-d3kx7-du4wt-r2fo6-uwc5a-hwhkk-m7snf-nxfhu-6bhgb-k6dn5-wae",
        },
        {
          "__principal__": "entn3-mas6a-37smu-vadtg-wgsno-zokif-vyphu-umase-lfwqe-dmcrk-kae",
        },
      ]
    `);
  });
});

describe('controllers', () => {
  it('should return the controllers of a canister with multiple controllers', async () => {
    const agent = new HttpAgent({ host: 'https://icp-api.io' });
    const status = await CanisterStatus.request({
      // Whoami canister
      canisterId: Principal.from('ivcos-eqaaa-aaaab-qablq-cai'),
      agent: agent,
      paths: ['controllers'],
    });
    expect((status.get('controllers') as Principal[]).map(p => p.toText())).toMatchInlineSnapshot(`
    [
      "hgfyw-myaaa-aaaab-qaaoa-cai",
      "b73qn-rqaaa-aaaap-aazvq-cai",
      "aux4w-bi6yf-a3bhr-zydhx-qvf2p-ymdeg-ddvg6-gmobi-ct4dk-wf4xd-nae",
      "jhnlf-yu2dz-v7beb-c77gl-76tj7-shaqo-5qfvi-htvel-gzamb-bvzx6-yqe",
    ]
  `);
  });
  it('should return the controllers of a canister with one controller', async () => {
    const agent = new HttpAgent({ host: 'https://icp-api.io' });
    const status = await CanisterStatus.request({
      // NNS Governance Canister
      canisterId: Principal.from('rrkah-fqaaa-aaaaa-aaaaq-cai'),
      agent: agent,
      paths: ['controllers'],
    });
    // Should be root canister
    expect((status.get('controllers') as Principal[]).map(p => p.toText())).toMatchInlineSnapshot(`
    [
      "r7inp-6aaaa-aaaaa-aaabq-cai",
    ]
  `);
  });
  it('should return the controllers of a canister with no controllers', async () => {
    const agent = new HttpAgent({ host: 'https://icp-api.io' });
    const status = await CanisterStatus.request({
      // nomeata's capture-the-ic-token canister
      canisterId: Principal.from('fj6bh-taaaa-aaaab-qaacq-cai'),
      agent: agent,
      paths: ['controllers'],
    });
    expect((status.get('controllers') as Principal[]).map(p => p.toText())).toMatchInlineSnapshot(`
    []
  `);
  });
});
