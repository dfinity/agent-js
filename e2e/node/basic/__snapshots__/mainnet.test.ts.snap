// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`certified query > should verify lots of query certificates > whoami seed 0 1`] = `"535yc-uxytb-gfk7h-tny7p-vjkoe-i4krp-3qmcl-uqfgr-cpgej-yqtjq-rqe"`;

exports[`certified query > should verify lots of query certificates > whoami seed 1 1`] = `"wf3fv-4c4nr-7ks2b-xa4u7-kf3no-32glf-lf7e4-4ng4a-wwtlu-a2vnq-nae"`;

exports[`certified query > should verify lots of query certificates > whoami seed 2 1`] = `"52mr2-fw2ng-2ofst-7jekz-xbymo-3ysz7-itwdk-bgstz-r7g4g-oz5vi-pqe"`;

exports[`certified query > should verify lots of query certificates > whoami seed 3 1`] = `"skpwg-42fe4-eyep5-nfyz7-66wvg-hthea-q3eek-vonbv-5wpxs-nxhmh-fqe"`;

exports[`certified query > should verify lots of query certificates > whoami seed 4 1`] = `"ghaya-cncjm-ntxgt-af5pp-6hzsz-tvwlv-hrlfc-ocq3t-ai7vk-vyixr-cqe"`;

exports[`certified query > should verify lots of query certificates > whoami seed 5 1`] = `"ebtho-zkeqd-ebs74-f5if3-lk6js-3a5q5-37xt4-3ylns-h7r4n-tkhly-aqe"`;

exports[`certified query > should verify lots of query certificates > whoami seed 6 1`] = `"5zqn5-627q2-spx2w-mt6bm-llsnz-gipvv-5femn-3box4-vzuh6-bn723-sae"`;

exports[`certified query > should verify lots of query certificates > whoami seed 7 1`] = `"tek7g-2zmny-nzjwg-ansf7-rkxv6-z32x6-3flbb-ous5d-pygjx-wkhlc-jae"`;

exports[`certified query > should verify lots of query certificates > whoami seed 8 1`] = `"jjn6g-sh75l-r3cxb-wxrkl-frqld-6p6qq-d4ato-wske5-op7s5-n566f-bqe"`;

exports[`certified query > should verify lots of query certificates > whoami seed 9 1`] = `"447dk-byguq-fqkfn-7h4r6-lpk74-itbnh-mkutj-m6tmy-igaff-hmfel-cqe"`;

exports[`certified query > should verify lots of query certificates > whoami seed 10 1`] = `"muo3f-ines5-bxbwm-6wi5e-z663m-3zte2-r7d4x-pleey-xqvxt-scwc5-jae"`;

exports[`certified query > should verify lots of query certificates > whoami seed 11 1`] = `"snzff-yj2qd-fjns7-lqhvw-rsgq7-tohk2-fjnw4-uq3d6-wtk56-pxgry-mqe"`;

exports[`certified query > should verify lots of query certificates > whoami seed 12 1`] = `"pb54o-aqais-24v7j-msopl-bqeuv-paefp-vuoqc-gkezk-grujb-oetl6-sae"`;

exports[`certified query > should verify lots of query certificates > whoami seed 13 1`] = `"bf37n-7ybos-wmqt6-yiov5-24q4m-ajpjk-madkr-snuj2-ngruk-tspnh-aqe"`;

exports[`certified query > should verify lots of query certificates > whoami seed 14 1`] = `"tbjmy-vay5e-hqvv6-sval4-zfxmm-aii6f-b7p55-hwtz5-dejtl-qcuh2-aqe"`;

exports[`certified query > should verify lots of query certificates > whoami seed 15 1`] = `"7d3pe-dh4ov-fp5xz-nctjc-5rduh-gzv3t-5ioyh-4dvx3-x4lgj-hz63t-dqe"`;

exports[`certified query > should verify lots of query certificates > whoami seed 16 1`] = `"37axv-sazcg-75pi3-owhxr-kollq-xnzjz-zfxsv-nzdbp-yaelp-shcul-jae"`;

exports[`certified query > should verify lots of query certificates > whoami seed 17 1`] = `"r772c-4dz5f-rpg4e-qzxgg-7bxlb-67zpu-bitgb-vsx7k-mmagd-6zk3d-4qe"`;

exports[`certified query > should verify lots of query certificates > whoami seed 18 1`] = `"knkon-d3kx7-du4wt-r2fo6-uwc5a-hwhkk-m7snf-nxfhu-6bhgb-k6dn5-wae"`;

exports[`certified query > should verify lots of query certificates > whoami seed 19 1`] = `"entn3-mas6a-37smu-vadtg-wgsno-zokif-vyphu-umase-lfwqe-dmcrk-kae"`;
